ecosystem/sep-0006.md

```diff
@@ -276,32 +276,82 @@ Example:
 
 Response code: `403 Forbidden`
 
-An anchor that requires the user to fill out information on a webpage hosted by the anchor should use this response. A wallet that receives this response should open a popup browser window to the specified URL.
+An anchor that requires the user to fill out information on a webpage hosted by the anchor should use this response. This can happen in situations where the anchor needs KYC information about a user, or when the anchor needs the user to perform a custom step for each transaction like entering an SMS code to confirm a withdrawal or selecting a bank account. A wallet that receives this response should open a popup browser window to the specified URL. Next, the wallet must do one of two things:
 
-The response body should be a JSON object with the following fields:
+1. If the transaction is a deposit and the response does not contain `interactive_deposit` set to `true`, then the wallet must make a second request to `/deposit` with the same parameters to receive the deposit information and show it to the user. This is used in the case of an asynchronous deposit, where the user will initiate the deposit through some external system like bitcoin or a bank transfer at some later time. The user could initiate zero, one, or many deposits using the deposit information shown.
+1. Otherwise, the wallet must either listen for a callback or poll the [`/transaction`](#single-historical-transaction) endpoint for updates about the transaction from the anchor. This is used for synchronous deposits or withdrawals. For synchronous deposits initiated in the anchor's flow, this allows the wallet to show the user status information and confirm if the deposit attempt initiated successfully or failed. For withdrawals, the wallet needs to get information on where to send the withdrawal payment to the anchor.
+
+The response body must be a JSON object with the following fields:
 
 Name | Type | Description
 -----|------|------------
 `type` | string | Always set to `interactive_customer_info_needed`
 `url` | string | URL hosted by the anchor. The wallet should show this URL to the user either as a popup or an iframe.
 `interactive_deposit` | boolean | (optional) Flag indicating that depositing is also handled in the anchor's interactive customer info flow. The wallet need not make additional requests to `/deposit` to complete the deposit. Defaults to `false`. Only relevant for responses to `/deposit` requests.
+`id` | string | (optional) The anchor's internal ID for this deposit / withdrawal request. Can be passed to the [`/transaction`](#single-historical-transaction) endpoint to check status of the request.
```

Comment:
I love this `id` field – it's definitely something that will enable us to track transactions more easily once they have moved from the `/deposit` or `/withdraw` flows.

My question is: should we be more explicit about when this `id` needs to be generated by the anchor? Is it only when the interactive KYC is completed? Or should an `id` be generated every time the endpoint is hit? 

Here's a suggestion, which might make the wallet-anchor integration even simpler:

1. We generate an `id` for each `/deposit` or `/withdraw` request, regardless of it being interactive or not
2. On future calls to `/deposit` or `/withdraw`, the wallet also provides an `id` query param – this will remove the need for the Wallet to always replicate all request params (+ newly required data) until it gets a success response, since `/deposit` is not stateless anymore
3. As soon as a 200 is received, the status of that deposit or withdrawal can be checked via `/transaction` by passing along that `id`

I think my reasoning is that we're basically doing a REST API around the `/transactions` resource, but using `GET /withdraw` or `GET /deposit` instead of `POST /transactions`, so it might make sense for them to share the same id.

Reply:
Thanks Alex, this makes sense. It's definitely a bummer that we need to be as backwards-compatible as possible, there are some strange choices in the existing SEP.

Here's what I'm thinking we can do for your suggestion:

1. Add an optional `id` parameter to `/deposit` and `/withdraw` success responses. These are useful for calling `GET /transaction`. They don't help with repeated calls to `/deposit` or `/withdraw`, because clients won't make any more of those requests once they get a success response.
1. After receiving a `interactive_customer_info_needed` response, the `interactive_deposit: false` case is the only one where the client needs to make another request to `/deposit`. We can specify that if the server has included an `id`, then the client doesn't need to re-include any parameters on its second call to `/deposit`, it just needs to supply the `id`.
1. Add an optional `id` parameter to the  `non_interactive_customer_info_needed` response, and specify that the client doesn't need to re-include any parameters if it supplies the `id` to future requests to `/deposit` or `/withdraw`.

Backwards compatibility analysis:

- old client and new server: may not work, client will need to be updated to check for `id` if server depends on statefulness
- new client and old server: will work since new clients check for `id`

However, I'm not too worried about this, because I don't think any existing flows are using the repeated requests to `/deposit` or `/withdraw` yet.

What do you think?

Reply:
This approach makes sense to me! Thanks :)